contracts/external/cw-vesting/src/contract.rs

```diff
@@ -0,0 +1,549 @@
+        return Err(ContractError::NoFundsToClaim);
+    }
+
+    // Update Vesting Payment with claimed amount and the correct status
+    VESTING_PAYMENT.update(deps.storage, |mut vp| -> Result<_, ContractError> {
+        // Decrease vesting amount
+        vp.amount = vp.amount.checked_sub(vested_amount)?;
+        // Increase the claimed amount
+        vp.claimed_amount = vp.claimed_amount.checked_add(vested_amount)?;
+
+        // Check if canceled and unbonding
+        if vp.status == VestingPaymentStatus::CanceledAndUnbonding {
+            // Set status to canceled as funds have unbonded if this executes
+            vp.status = VestingPaymentStatus::Canceled
+        }
+
+        // If the amount remaining in contract goes to zero, update status
+        if vp.amount == Uint128::zero() {
+            vp.status = VestingPaymentStatus::FullyVested
+        }
+        Ok(vp)
+    })?;
+
+    let mut msgs: Vec<CosmosMsg> = vec![];
+
+    // Get transfer message for recipient payout
+    if vested_amount != Uint128::zero() {
+        let transfer_msg = vesting_payment
+            .denom
+            .get_transfer_to_message(&vesting_payment.recipient, vested_amount)?;
+        msgs.push(transfer_msg);
+    }
+
+    // Get owner's payout message in event of canceled and unbonding
+    if vesting_payment.status == VestingPaymentStatus::CanceledAndUnbonding {
+        let unvested_amount = vesting_payment.amount.checked_sub(vested_amount)?;
+        let owner = cw_ownable::get_ownership(deps.storage)?;
+        if let Some(owner) = owner.owner {
+            if unvested_amount != Uint128::zero() {
+                let withdraw_unbonded_msg = vesting_payment
+                    .denom
+                    .get_transfer_to_message(&owner, unvested_amount)?;
+                msgs.push(withdraw_unbonded_msg);
+            }
+        }
+    }
+
+    Ok(Response::new()
+        .add_attribute("method", "distribute")
+        .add_attribute("vested_amount", vested_amount)
+        .add_messages(msgs))
+}
+
+pub fn execute_update_owner(
+    deps: DepsMut,
+    info: MessageInfo,
+    env: Env,
+    action: cw_ownable::Action,
+) -> Result<Response, ContractError> {
+    let ownership = cw_ownable::update_ownership(deps, &env.block, &info.sender, action)?;
+    Ok(Response::default().add_attributes(ownership.into_attributes()))
+}
+
+pub fn execute_delegate(
+    _env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    validator: String,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    // Non-payable as this method delegates funds in the contract
+    nonpayable(&info)?;
+
+    let vesting_payment = VESTING_PAYMENT.load(deps.storage)?;
+
+    // Check status is active
+    if vesting_payment.status != VestingPaymentStatus::Active {
+        return Err(ContractError::NotActive);
+    }
+
+    // Check sender is the recipient of the vesting payment
+    if info.sender != vesting_payment.recipient {
```

Comment:
nit: it's nice to have these authorization checks happen first as it delivers the most relevant / restrictive error first

Reply:
Feel free to make a commit or PR.